### PR TITLE
grad_mode decorators without paren

### LINKED ISSUE #107036: torch.no_grad(fn)
### 🚀 The feature, motivation and pitch

I think it *could* be nice to have the following:

```python
import torch

def no_grad(fn: Callable) -> Callable:
    @torch.no_grad()
    def no_grad_fn(*args, **kwargs):
        return fn(*args, **kwargs)
    return no_grad_fn

def square(x):
    return x**2

square_no_grad = no_grad(square)

if __name__ == "__main__":
    x = torch.tensor([1.0, 2.0, 3.0], requires_grad=True)

    print(square(x))
    # tensor([1., 4., 9.], grad_fn=<PowBackward0>)
    print(square_no_grad(x))
    # tensor([1., 4., 9.])
```

It could be nice to have the `no_grad` be a torch function --> `torch.no_grad(fn)`.

Why? Consider a scenario where we have a core function call. Sometimes we want to have it track gradients, but in other contexts we especially don't. Instead of copying the code twice and decorating one with `@torch.no_grad()`, this seems like a cleaner way.

However, I'm not sure there are so many use cases for this? I have one myself but I don't know how common it will be.

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by andreasfloros**
This PR implements the feature described in #107036 for `no_grad`, `enable_grad` and `inference_mode`.

Users can still use the above as before but they can also use them without parentheses.

For example:

```python
import torch

a = torch.ones(1, requires_grad=True)


def do_something():
    print(2 * a)


with torch.no_grad():
    do_something()  # tensor([2.])

torch.no_grad()(do_something)()  # tensor([2.])

torch.no_grad(do_something)()  # tensor([2.])

do_something()  # tensor([2.], grad_fn=<MulBackward0>)
```

For `inference_mode`, decorating without parenthesis is equivalent to decorating with the default `mode=True`, similiar to how dataclasses behave (https://docs.python.org/3/library/dataclasses.html#module-contents)

Closes #107036

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by andreasfloros** [comment]
The merge failed due to a mypy errors:
``__new__" must return a class instance (got "Union[_NoParamDecoratorContextManager, F]") ``

I've removed the `__new__` type annotations since they weren't being picked up anyway.

**Comment by kit1980** [comment]
@andreasfloros the lint is still failing...

**Comment by andreasfloros** [comment]
> @andreasfloros the lint is still failing...

Sorry about that, missed a ,. Should be fine now.

@pytorchbot  merge

**Comment by kit1980** [comment]
@andreasfloros because you're a first time contributor, every time you update the code a project member needs to approve the CI run.

**Comment by andreasfloros** [comment]
@kit1980  I see, thanks for helping me out!
